### ISSUE TITLE
[metal] Use of undeclared identifier in functions

### ISSUE THREAD

**Issue by dati91** [KhronosGroup/SPIRV-Cross#610]
I have the following shader: [glsl shader](https://gist.github.com/dati91/c394ecada63eef768e78983fd813bc8a)
And the generated metal shader: [metal shader](https://gist.github.com/dati91/1990eb6c04fc8283dcceedb3872c3d2f)
And the problem with it: [compilation errors](https://gist.github.com/dati91/5fa1dcc940d9bebf1ac01e5659a372dd)

In some functions, we use global variables like `gl_Position`. In the generated code those are defined in an out structure, and never passed to the functions.

**Comment by HansKristian-ARM** [comment]
Confirmed, MSL fails to forward gl_Position to child function, it just forwards the block itself, which is weird.

**Comment by HansKristian-ARM** [comment]
Trivial repro:
```
#version 450

void write_position()
{
	gl_Position = vec4(1.0);
}

void main()
{
	write_position();
}
```
becomes:
```
#pragma clang diagnostic ignored "-Wmissing-prototypes"

#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct main0_out
{
    float4 gl_Position [[position]];
};

void write_position(thread float4& v_15)
{
    gl_Position = float4(1.0);
}

vertex main0_out main0()
{
    main0_out out = {};
    write_position(v_15);
    return out;
}
```

**Comment by HansKristian-ARM** [comment]
@billhollings the issue is here:
https://github.com/KhronosGroup/SPIRV-Cross/blob/master/spirv_msl.cpp#L540

What happens is that the global variable is found to be accessed in the function, but it doesn't check that children of that struct could be builtins, which are handled specially in MSL, so it tries to forward the block itself. We might need some special handing of Blocks which include builtins. Is this something you could look at?

**Comment by billhollings** [comment]
@HansKristian-ARM 

Sure. I'll have a look.

**Comment by billhollings** [comment]
Fixed in PR #621.
